sqlalchemy-stubs/sql/sqltypes.pyi

```diff
@@ -3,78 +3,107 @@ from datetime import datetime
+from ..engine import Engine
+from ..util import langhelpers
 
 _T = TypeVar("_T")
-
-class _LookupExpressionAdapter:
-    class Comparator(TypeEngine.Comparator): ...
-    comparator_factory: Any = ...
-
-class Concatenable:
-    class Comparator(TypeEngine.Comparator): ...
-    comparator_factory: Any = ...
-
-class Indexable:
-    class Comparator(TypeEngine.Comparator):
-        def __getitem__(self, index: Any): ...
-    comparator_factory: Any = ...
-
-class String(Concatenable, TypeEngine[str]):
-    __visit_name__: str = ...
-    RETURNS_UNICODE: Any = ...
-    RETURNS_BYTES: Any = ...
-    RETURNS_CONDITIONAL: Any = ...
-    RETURNS_UNKNOWN: Any = ...
-    length: Any = ...
-    collation: Any = ...
+_TE = TypeVar("_TE", bound=TypeEngine[Any])
+_ST = TypeVar("_ST", bound=SchemaType)
+_EN = TypeVar("_EN", bound=Enum)
+
+_NumericType = Union[float, Decimal]
+_JSONType = Union[str, Mapping[Any, Any], List[Any]]
+
+class _LookupExpressionAdapter(Generic[_T]):
+    class Comparator(TypeEngine.Comparator[_TE]): ...
+    comparator_factory: Type[Comparator[TypeEngine[_T]]] = ...
+
+class Concatenable(Generic[_T]):
+    class Comparator(TypeEngine.Comparator[_TE]): ...
+    comparator_factory: Type[Comparator[TypeEngine[_T]]] = ...
+
+class Indexable(Generic[_T]):
+    class Comparator(TypeEngine.Comparator[_TE]):
+        def __getitem__(self, index: Any) -> Any: ...
+    comparator_factory: Type[Comparator[TypeEngine[_T]]] = ...
+
+class String(Concatenable[str], TypeEngine[str]):  # type: ignore[misc]
+    __visit_name__: str = ...
+    RETURNS_UNICODE: ClassVar[langhelpers._symbol] = ...
+    RETURNS_BYTES: ClassVar[langhelpers._symbol] = ...
+    RETURNS_CONDITIONAL: ClassVar[langhelpers._symbol] = ...
+    RETURNS_UNKNOWN: ClassVar[langhelpers._symbol] = ...
+    length: Optional[int] = ...
+    collation: Optional[str] = ...
+    @overload
     def __init__(
         self,
-        length: Optional[Any] = ...,
-        collation: Optional[Any] = ...,
+        length: Optional[int] = ...,
+        collation: Optional[str] = ...,
         convert_unicode: bool = ...,
-        unicode_error: Optional[Any] = ...,
         _warn_on_bytestring: bool = ...,
         _expect_unicode: bool = ...,
     ) -> None: ...
-    def literal_processor(self, dialect: Any): ...
-    def bind_processor(self, dialect: Any): ...
-    def result_processor(self, dialect: Any, coltype: Any): ...
+    @overload
+    def __init__(
+        self,
+        length: Optional[int],
+        collation: Optional[str],
+        convert_unicode: Literal["force"],
+        unicode_error: str,
+        _warn_on_bytestring: bool = ...,
+        _expect_unicode: bool = ...,
+    ) -> None: ...
+    @overload
+    def __init__(
+        self,
+        length: Optional[int],
+        *,
+        convert_unicode: Literal["force"],
+        unicode_error: str,
+        _warn_on_bytestring: bool = ...,
+        _expect_unicode: bool = ...,
+    ) -> None: ...
```

Comment:
how is this overload different from the above one?

Reply:
The above overload handles positional arguments (`String(40, '...', 'force', '...')`) and this overload handles named parameters (`String(convert_unicode='force', unicode_error='...')`). I noticed a problem in how I wrote them, so I've pushed a commit to fix it. However, this is really the only way to make `unicode_error` required when `convert_unicode` is `'force'`.

Reply:
ok I guess. They still seem the same to me, but I trust your judgment. 👍